lib/puppet/module_tool/tar/gnu.rb

```diff
@@ -4,15 +4,14 @@
 
 class Puppet::ModuleTool::Tar::Gnu
   def unpack(sourcefile, destdir, owner)
-    sourcefile = File.expand_path(sourcefile)
+    safe_sourcefile = Shellwords.shellescape(File.expand_path(sourcefile))
     destdir = File.expand_path(destdir)
+    safe_destdir = Shellwords.shellescape(destdir)
 
-    Dir.chdir(destdir) do
-      Puppet::Util::Execution.execute("gzip -dc #{Shellwords.shellescape(sourcefile)} | tar xof -")
-      Puppet::Util::Execution.execute("find . -type d -exec chmod 755 {} +")
-      Puppet::Util::Execution.execute("find . -type f -exec chmod u+rw,g+r,a-st {} +")
-      Puppet::Util::Execution.execute("chown -R #{owner} .")
-    end
+    Puppet::Util::Execution.execute("gzip -dc #{safe_sourcefile} | tar --extract --no-same-owner --directory #{safe_destdir} --file -")
```

Comment:
If this is just for GNU tar then can't this be converted to avoid a pipe (and thus a shell):
```suggestion
    Puppet::Util::Execution.execute(['tar', '--extract', '--no-same-owner', '--gzip', '--directory', destdir, '--file', sourcefile])
```

Reply:
I agree we should make the change you recommended. The only hiccup is the `puppet module install` command uses the following to detect which tar implementation to use:

https://github.com/puppetlabs/puppet/blob/56d07e923c9c7cfd1a9c4e8553ae30fcbd82ee25/lib/puppet/module_tool/tar.rb#L11-L15

On Windows and Solaris, we package the minitar gem

https://github.com/puppetlabs/puppet-runtime/blob/967911be9d599aa2d18cb0ba2699adb23cd89aa2/configs/projects/_shared-agent-components.rb#L64-L65

So I think the change you're proposing would work for everything else except for AIX. Though we should probably just package the minitar gem on AIX and keep your proposed changes.

Reply:
I assumed there would be a posix tar implementation, besides gnu. I don't know AIX and whether it's better to use posix compatible options or depend on a gem

Reply:
There's `gtar`, but it's not installed by default. For example https://www.ibm.com/support/pages/how-install-gnu-tar-aix-extract-spectrum-control-installation-media

Reply:
IIRC BSD also ships a different tar implementation which may or may not support those options. https://stackoverflow.com/questions/37475833/does-tar-follow-either-the-posix-and-or-gnu-command-line-options-conventions/37477473#37477473 says POSIX didn't standardize `tar`, which doesn't make this easier.

Reply:
I'm thinking the safest thing right now is to stick with shelling out and escaping args.

Reply:
I agree. That's why I made 2 separate review comments